docs/TDD.md

```diff
@@ -0,0 +1,220 @@
+
+If one or more tasks remain failing, then the above process is repeated with each one until all tests pass.
+
+Sometimes, the expected data and returned data are too large to all be placed in the `Test Failure` section.
+It may look something like this:
+
+```
+AssertionError: '("Sc[67 chars]\')\n\n(\'Brass Spyglass\', \'Abandoned Lighth[952 chars]')\n' != '("Sc[67 chars]\')\n(\'Brass Spyglass\', \'Abandoned Lighthou[928 chars]')\n'
+Diff is 970 characters long. Set self.maxDiff to None to see it.
+```
+
+There may still be enough data to see what the problem is.
+In the case above, there are two line feeds returned (e.g. `\n\n(\'Brass Spyglass`) when only one is expected (e.g. `\n(\'Brass Spyglass`).
+
+## After all the tests pass
+
+Congratulations!
+All the tests have passed.
+What next?
+The solution could be published right away.
+Or, now that the code works, if you would like to refactor it for any reason, you can modify the code and submit another iteration.
+If you think the code could be better, but you don't know how, you can request mentoring for the solution.
+If a mentor is available with ideas for other approaches to a solution, they may contact you to share those ideas.
+You could allow comments when publishing your solution, and other students may take the opportunity to make comments or ask questions.
+
+## Optimizing for Performance
+
+Although "Premature optimization is the root of all evil" (a saying attributed to both Tony Hoare and Donald Knuth), there does come a time when the solution works and it is desired to make it as performant as possible.
+One of those times may be when the solution passes some tests but times out for others.
+It can be helpful to know exactly how much time a piece of code is taking.
+The [timeit][timeit] module can be used to time the execution of code down to very small durations.
+The `timeit` function can take up to five arguments.
+The `stmt` parameter defines the actual code to be run and timed.
+The `stmt` code will be run for a specifed `number` of iterations.
+The `setup` parameter defines the code that is run only once to prepare for running the `stmt` code.
+The time for the `setup` code to run is included in the overall time.
+The more iterations the `stmt` code is run, the less the `setup` time should count per iteration.
+The `timer` parameter allows for passing in a different `Timer` from the default.
+The default argument for the `timer` parameter is [perf_counter][perf_counter], which should suffice.
+The `number` parameter defines the number of times the `stmt` code will be run.
+The default argument for the `number` parameter is `1_000_000`.
+The `globals` parameter specifies a namespace in which to execute the code.
+The default argument for the `globals` parameter is `None`.
+
+Following is an example of using `timeit` to see how long it takes to determine if a sentence contains all English vowels:
```

Comment:
```suggestion
The following is an example of using `timeit` to see how long it takes to determine if a sentence contains all English vowels:
```

Optional, this is how I often write, but I know that it is not correct.

Reply:
I was about to query why it's not correct, but then I looked it up and now I see that "The following" refers to some thing, whereas "Following" refers to a point in time. So this correction makes sense in a strict sense. [Following vs. The Following](https://dictionary.cambridge.org/us/grammar/british-grammar/following-or-the-following).